ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,436 @@
+	}
+	size_t high_part_shift = POW_10_LUT[BC_VECTOR_SIZE - n2_top_digits + 1];
+	size_t low_part_shift = POW_10_LUT[n2_top_digits - 1];
+	BC_VECTOR divider = n2_vector[n2_top_index] * high_part_shift + n2_vector[n2_top_index - 1] / low_part_shift;
+	for (i = 0; i < quot_arr_size; i++) {
+		BC_VECTOR divided = n1_vector[n1_top_index - i] * high_part_shift + n1_vector[n1_top_index - i - 1] / low_part_shift;
+
+		/* If it is clear that n2 is greater in this loop, then the quotient is 0. */
+		if (div_carry == 0 && divided < divider) {
+			quot_vector[quot_top_index - i] = 0;
+			div_carry = n1_vector[n1_top_index - i];
+			n1_vector[n1_top_index - i] = 0;
+			continue;
+		}
 
-			if (carry != 0) {
-				*rptr = carry;
+		/*
+		 * Determine the temporary quotient.
+		 * "divided" is n1_high in the previous example. The maximum value of n1_high is n2_high * B,
+		 * so here it is divider * BC_VECTOR_BOUNDARY_NUM.
+		 * The number of digits in divider is BC_VECTOR_SIZE + 1, so even if divider is at its maximum value,
+		 * it will never overflow here.
+		 */
+		divided += div_carry * BC_VECTOR_BOUNDARY_NUM * high_part_shift;
+		BC_VECTOR quot_guess = divided / divider;
+
+		/* For sub, add the remainder to the high-order digit */
+		n1_vector[n1_top_index - i] += div_carry * BC_VECTOR_BOUNDARY_NUM;
+
+		/*
+		 * The temporary quotient can only have errors in the "too large" direction.
+		 * So if the temporary quotient is 0 here, the quotient is 0.
+		 */
+		if (quot_guess == 0) {
+			quot_vector[quot_top_index - i] = 0;
+			div_carry = n1_vector[n1_top_index - i];
+			n1_vector[n1_top_index - i] = 0;
+			continue;
+		}
+
+		/* sub */
+		BC_VECTOR sub;
+		BC_VECTOR borrow = 0;
+		BC_VECTOR *n1_calc_bottom = n1_vector + n1_arr_size - n2_arr_size - i;
+		for (j = 0; j < n2_arr_size - 1; j++) {
+			sub = n2_vector[j] * quot_guess + borrow;
+			BC_VECTOR sub_low = sub % BC_VECTOR_BOUNDARY_NUM;
+			borrow = sub / BC_VECTOR_BOUNDARY_NUM;
+
+			if (n1_calc_bottom[j] >= sub_low) {
+				n1_calc_bottom[j] -= sub_low;
+			} else {
+				n1_calc_bottom[j] += BC_VECTOR_BOUNDARY_NUM - sub_low;
+				borrow++;
 			}
 		}
+		/* last digit sub */
+		sub = n2_vector[j] * quot_guess + borrow;
+		bool neg_remainder = n1_calc_bottom[j] < sub;
+		n1_calc_bottom[j] -= sub;
+
+		/* If the temporary quotient is too large, add back n2 */
+		BC_VECTOR carry = 0;
+		if (neg_remainder) {
+			quot_guess--;
+			for (j = 0; j < n2_arr_size - 1; j++) {
+				n1_calc_bottom[j] += n2_vector[j] + carry;
+				carry = n1_calc_bottom[j] / BC_VECTOR_BOUNDARY_NUM;
+				n1_calc_bottom[j] %= BC_VECTOR_BOUNDARY_NUM;
+			}
+			/* last add */
+			n1_calc_bottom[j] += n2_vector[j] + carry;
+		}
+
+		quot_vector[quot_top_index - i] = quot_guess;
+		div_carry = n1_vector[n1_top_index - i];
+		n1_vector[n1_top_index - i] = 0;
 	}
 }
 
+static void bc_do_div(char *n1, size_t n1_readable_len, size_t n1_bottom_extension, char *n2, size_t n2_len, bc_num *quot, size_t quot_len)
```

Comment:
What is the "bottom extension" ?

Reply:
For calculations that specify a scale larger than the scale of n1, the scale of n1 must be expanded and padded with zeros before the calculation can be performed.

This refers to the number of digits of scale extension.

Reply:
Please add a comment before the function to explain this.

Also should the `char*` be const or not?

Reply:
I think it can be made const. I'll try to fix it.

Reply:
Fixed 
https://github.com/php/php-src/pull/14660/commits/93e34b77211096a9d44539ff8aa658f58284e264